RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md

```diff
@@ -0,0 +1,301 @@
+Transform: AWS::LanguageExtensions
+Parameters:
+  Country:
+    Type: String
+Mappings:
+  DNS:
+    usa:
+      dns: mypage.com
+      ttl: '600'
+    canada:
+      dns: mypage.ca
+      ttl: '600'
+    germany:
+      dns: mypage.de
+      ttl: '300'
+Resources:
+  DemoApiGateway:
+    Type: 'AWS::ApiGatewayV2::Api'
+    Properties:
+      Name: Demo API Gateway
+      ProtocolType: HTTP
+  DemoApiStage:
+    Type: 'AWS::ApiGatewayV2::Stage'
+    Properties:
+      ApiId: !Ref DemoApiGateway
+      StageName: live
+      Description: Live Stage
+      AutoDeploy: true
+  DemoApiDomainName:
+    Type: 'AWS::ApiGatewayV2::DomainName'
+    Properties:
+      DomainName: !FindInMap 
+        - DNS
+        - !Ref country
+        - dns
+        - DefaultValue: !FindInMap 
+            - DNS
+            - usa
+            - dns
+```
+
+With default value added, the size of Mapping section is significantly reduced. In addition, there is no need to update template when new countries that uses default DNS are added.
+
+#### Example 2
+
+Here is another example using default value to set EC2 instance type
+
+```
+AWSTemplateFormatVersion: 2010-09-09
+Transform: AWS::LanguageExtensions
+Mapping:
+  InstanceConfiguration:
+    us-east-1:
+      Type: m5.large
+    us-west-2:
+      Type: m5.large
+    eu-west-1:
+      Type: m5.medium
+Resources:
+  Instance:
+    Type: 'AWS::EC2::Instance'
+    Properties:
+      InstanceType: !FindInMap 
+        - 'InstanceConfiguration'
+        - !Ref 'AWS::Region'
+        - 'Type'
+        - DefaultValue: m5.small
+# Use m5.small instance unless it's in us-east-1, us-west-2 or eu-west-1
+```
+
+## Limitation
+
+There will be few limitations when using other intrinsic functions or default value in `Fn::FindInMap`:
+
+1. User can not use `Fn::Split` to supply argument list (partial or whole) of `Fn::FindInMap`. However, please do note that it **CAN** be used in conjunction with other functions that produce a `string` such as `Fn::Select`. Here is an example:
+
+```
+# This is not allowed
+!FindInMap !Split [".", "RegionMap.us-east-1.HVM64"] ]
+
+# This is allowed
+!FindInMap
+  - !Select [0, !Split [ ".", "RegionMap.us-east-1.HVM64" ] ]
+  - !Select [1, !Split [ ".", "RegionMap.us-east-1.HVM64" ] ]
+  - !Select [2, !Split [ ".", "RegionMap.us-east-1.HVM64" ] ]
+```
+
+1. `Fn::FindInMap` enhancement is made available via [`AWS::LanguageExtensions` transform](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/transform-aws-languageextensions.html), hence `Fn::FindInMap` can only support operations whose return values are known during transform time. When unsupported intrinsic functions are used, an exception will be returned back to customers.
+    1. Here are examples not supported:
+        1. `Ref` to a resource.
+        2. `Fn::GetAtt` to retrieve an attribute from a resource.
+        3. `Fn::Sub` with nested `Fn::GetAtt` or `Ref` which points to a resource or resource attribute. 
+
+```
+Mapping:
```

Comment:
Add a note above this example saying something like. "For example, the following will NOT be supported:" and then add a comment in the YAML just above each `!GetAtt LoadBalancer.DNSName` line saying "GetAtt of a resource attribute not supported by DefaultValue" to make it super clear that this is an example of what will NOT be supported.